### PR TITLE
Eg-304, refactor Text, TextItems etc to allow out of order translations

### PR DESCRIPTION
**Pull request by cmrhmrc**

### CODE REVIEW

File: app/models/ui/UIComponent.scala [hunk 1]
```diff
@@ -20,35 +20,41 @@ import play.api.i18n.Lang
...
+  def value(implicit lang: Lang): Seq[TextItem] = if (lang.code.equals("cy")) welsh else english
+  //def isEmpty: Boolean = english.isEmpty
+  def isEmpty(implicit lang: Lang): Boolean = if (lang.code.equals("cy")) welsh.isEmpty else english.isEmpty
+  def toWords(implicit lang: Lang): Seq[String] = value(lang).flatMap(_.toWords)
+  override def toString: String = s"[${english.map(t => t.toString).mkString("")}:${welsh.map(t => t.toString).mkString("")}]"
+  def +(other: Text): Text = Text(english ++ other.english, welsh ++ other.welsh)
+}
+
+object Text {
+  def apply(phrase: Vector[String]): Text = Text(Words(phrase(0)), Words(phrase(1)))
```
Comment (woollett-r) [inline app/models/ui/UIComponent.scala]: Is apply for Phrase correct? phrase(0) and phrase(1) are basically strings like "english" and "welsh" for the other apply methods. However, the other methods use Seq(Words(String)) not Words(String).

File: app/models/ui/UIComponent.scala [hunk 2]
```diff
@@ -20,35 +20,41 @@ import play.api.i18n.Lang
...
 
-case class PageLink(dest: String, txt: Text) extends TextItem {
-  override def toString: String = s"[link:$txt:$dest]"
-  def isEmpty: Boolean = txt.isEmpty
+//
+// Bilingual
+//
+case class Text(english: Seq[TextItem], welsh: Seq[TextItem]) {
+  def value(implicit lang: Lang): Seq[TextItem] = if (lang.code.equals("cy")) welsh else english
+  //def isEmpty: Boolean = english.isEmpty
```
Comment (woollett-r) [inline app/models/ui/UIComponent.scala]: Remove earlier version of isEmpty

Comment (cmrhmrc) [reply app/models/ui/UIComponent.scala]: Remvoed this earlier today, you;ll need to pull the branch again

File: app/models/ui/UIComponent.scala (hunk 1 above)
Comment (cmrhmrc) [reply app/models/ui/UIComponent.scala]: Text(Words(phrase(0)), Words(phrase(1))) will match with Text(TextItem, TextItem) and create a Text(Seq(TextItem),Seq(TextItem)) as Words extends TextItem

File: test/models/ui/UIComponentsSpec.scala [hunk 3]
```diff
@@ -125,142 +135,114 @@ class UIComponentsSpec extends BaseSpec {
...
 
-      bulletPointList.listItems.head.head.toString mustBe "[Bullet point 1 text:Welsh bullet point 1 text:false]"
-      bulletPointList.listItems.head(1).toString mustBe "[link:[Bullet point 1 link:Welsh bullet point 1 link:false]:http://bulletPointOneUrl:false]"
+      bulletPointList.listItems.head.english(0).toString mustBe "Bullet point 1 text"
+      bulletPointList.listItems.head.english(1).toString mustBe "[link:Bullet point 1 link:http://bulletPointOneUrl:false]"
 
-      bulletPointList.listItems(1).head.toString mustBe "[Bullet point 2 text:Welsh bullet point 2 text:false]"
-      bulletPointList.listItems(1)(1).toString mustBe "[link:[Bullet point 2 link:Welsh bullet point 2 link:false]:http://bulletPointTwoUrl:true]"
+      bulletPointList.listItems(1).english(0).toString mustBe "Bullet point 2 text"
+      bulletPointList.listItems(1).english(1).toString mustBe "[link:Bullet point 2 link:http://bulletPointTwoUrl:true]"
```
Comment (woollett-r) [inline test/models/ui/UIComponentsSpec.scala]: This is probably a pre-existing problem, but at present we check only the english version of the bullet point list. For completeness I suggest we check the Welsh as well.

Comment (cmrhmrc) [reply test/models/ui/UIComponentsSpec.scala]: Yes, this reflects the original code

File: test/services/TextBuilderSpec.scala [hunk 4]
```diff
@@ -95,63 +121,63 @@ class TextBuilderSpec extends BaseSpec {
 
       val text = ""
 
-      TextBuilder.fragmentsToDisplay( text ) mustBe text
+      BulletPointBuilder.fragmentsToDisplay( text ) mustBe text
```
Comment (woollett-r) [inline test/services/TextBuilderSpec.scala]: The tests for BulletPointBuilder.fragmentsToDisplay should really be in BulletPointBuilderSpec.

Comment (cmrhmrc) [reply test/services/TextBuilderSpec.scala]: Yes agree, I'll move them and push to branch when done

File: app/views/components/link.scala.html [hunk 5]
```diff
@@ -14,11 +14,12 @@
  * limitations under the License.
  *@
 
-@import models.ui.HyperLink
+@import models.ui.Link
 
-@(link: HyperLink)(implicit messages: Messages)
+@(link: Link)(implicit messages: Messages)
```
Comment (woollett-r) [inline app/views/components/link.scala.html]: Do we still require the implicit messages here? Looking at render_text I suspect we will have either an english or welsh link passed into this component.

Comment (cmrhmrc) [reply app/views/components/link.scala.html]: No we dont, I will remove

File: app/views/components/page.scala.html [hunk 6]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  *@
 
-@import models.ui.{Page, Paragraph, HyperLink, H1, H2, H3, BulletPointList, Question}
+@import models.ui.{Page, Paragraph, Link, H1, H2, H3, BulletPointList, Question}
```
Comment (woollett-r) [inline app/views/components/page.scala.html]: Link is not actually used in this component.

Comment (cmrhmrc) [reply app/views/components/page.scala.html]: will remove

File: test/models/ui/UIComponentsSpec.scala (hunk 3 above)
Comment (woollett-r) [reply test/models/ui/UIComponentsSpec.scala]: I suggest we fix this when we fix the bullet point list bug!

**Comment by woollett-r** [comment]
I have finished my review. I think there are just a few minor issues to tidy up and the branch can be merged. However, I'll ask Mac if he wants to have a look as well.

**Comment by cmrhmrc** [comment]
> I have finished my review. I think there are just a few minor issues to tidy up and the branch can be merged. However, I'll ask Mac if he wants to have a look as well.

All updates complete and pushed to branch

**Pull request merged**